File: scala/sieve/src/main/scala/Sieve.scala
object Sieve {

  def sieve(remainder: List[Int], primes: List[Int]): List[Int] =
    remainder match {
      case Nil => primes.reverse
      case prime::xs => sieve(xs.filter(x => x % prime > 0), prime :: primes)
    }

  def primesUpTo(max: Int) = sieve(List.range(2, max + 1), Nil)
}

File: cpp/sieve/sieve.cpp
#include <vector>
#include "sieve.h"

namespace sieve
{
    std::vector<int> primes(int max)
    {
        std::vector<bool> sieve(max + 1, true);
        std::vector<int> primes;

        for (int i = 2; i <= max; i++)
        {
            if (!sieve[i])
                continue;

            primes.push_back(i);

            for (int j = i * 2; j <= max; j += i)
                sieve[j] = false;
        }

        return primes;
    }
} // namespace sieve
